### DISCUSSION TITLE
Using !expr in 'label:'  chunk option is not possible?

### DISCUSSION THREAD

**Discussion by r-scotti** [quarto-dev/quarto-cli#6210] [Q&A] [answered]
Hello.
In the small test below the last chunk doesn't compile.
If I substitute the first line: <#| label: !expr taste0("tbl-", tab)>
with <#| label: tbl-tables>, the output is as expected.
Is it just not possible to use "!expr" in the 'Label:' chunk option?
Many thanks, Roberto

Details
quarto::quarto_version()    ‘0.9.238’  (how to upgrade?)
System: Ubuntu 20.04
> Sys.info()
  sysname:    "Linux"  
  release:  "5.15.0-76-generic" 
  version: "#83~20.04.1-Ubuntu SMP Wed Jun 21 20:23:31 UTC 2023"
Below: Example Quarto code

````qmd
---
format: pdf
---

```{r}
library(tidyverse)
library(kableExtra)
x <- c(3, 1, 2)
Table1 <- data.frame(A = x, B = letters[x])
x <- c(11, 16, 21)
Table2 <- data.frame(C = x, D = letters[x])

tab <- "Table1"
```

OK with "#| label: tbl-tables"

```{r}
#| label: !expr paste0("tbl-", tab)
#| tbl-cap: !expr paste("This is", tab)
#| tbl-subcap:
#|   - "first"
#|   - "second"

tbl <- get(tab)

kbl(tbl)

tibble(attribute = names(tbl)) |> 
  unnest(cols = c(attribute)) |> 
  mutate(kind = typeof(attribute)) |> 
  kbl()
```

**Comment by mcanouil** [comment]
Could you format properly your post? In its current shape it is not quite legible.
For instance don't use title for text.

See <https://quarto.org> for your question about upgrade.

---

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by r-scotti** [reply]
Excuse me, first time here. I edited, is it ok now?

**Reply by cscheid** [reply]
```
#| label: !eppr taste0("tbl-", tab)
```

You have at least two typos in your code here (`eppr`, and `taste0`). Please, do make sure to proofread these.

Specifically, we 100% need the code you post to match _exactly_ the code you're running, or there's no way for us to debug.

**Reply by r-scotti** [reply]
I apologise, hope it is ok now

**Reply by cscheid** [reply]
This isn't a quarto problem, it's a `knitr` one:

```
$ R
> knitr:::knit("6210.qmd")


processing file: 6210.qmd
  |..................                                                    |  25%
  ordinary text without R code

  |...................................                                   |  50%
label: unnamed-chunk-1
  |....................................................                  |  75%
  ordinary text without R code

  |......................................................................| 100%
Quitting from lines 19-33 (6210.qmd)
Error in defaults[[name]] : invalid subscript type 'list'
```

**Reply by mcanouil** [reply]
@r-scotti you should report this in `knitr` repository, see <https://github.com/yihui/knitr>.

**Reply by r-scotti** [reply, marked as answer]
Many thanks! Your redirection towards     knitr   has been the key to solve my problem.
I immediately referred to GitHub Kintr site where I found that an analogous question had already been answered.
I developed my solution based on https://stackoverflow.com/questions/12095113/r-knitr-possible-to-programmatically-modify-chunk-labels
Excuse me again for the errors.
Roberto

**Reply by cderv** [reply]
FWIW **knitr** does not support using a R object / expression as a label. 

This is because the labels are determined before any code chunk is evaluated (i.e., they are determined at the parsing stage): [yihui/knitr@0e9bb8b/R/parser.R#L197-L198](https://github.com/yihui/knitr/blob/0e9bb8bc8cd77b2b8ace872637bb86a5b6d54bcb/R/parser.R#L197-L198) When parsing the document, the parser doesn't know the value of any objects.

Related discussion: https://github.com/rstudio/bookdown/issues/1419#issuecomment-1516774393

**Comment by r-scotti** [comment]
Many thanks!
I could imagine but I do not know how I could find the answer you are
giving me.
While compiling the question in knitr GitHub site, a question
tightly related to mine came up and I discovered that using
knitr::knit_extend()  it is possible (though not straightforward at all!)
to obtain my goal:
"produce many cross-referenceable nicely formatted (with label, cap and
sub-caps) table pairs, writing the code only once."
Roberto

Il giorno lun 17 lug 2023 alle ore 15:40 Christophe Dervieux <
***@***.***> ha scritto:

> FWIW *knitr* does not support using a R object / expression as a label.
>
> This is because the labels are determined before any code chunk is
> evaluated (i.e., they are determined at the parsing stage):
> ***@***.***/R/parser.R#L197-L198
> <https://github.com/yihui/knitr/blob/0e9bb8bc8cd77b2b8ace872637bb86a5b6d54bcb/R/parser.R#L197-L198>
> When parsing the document, the parser doesn't know the value of any objects.
>
> Related discussion: rstudio/bookdown#1419 (comment)
> <https://github.com/rstudio/bookdown/issues/1419#issuecomment-1516774393>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/quarto-dev/quarto-cli/discussions/6210#discussioncomment-6467929>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEAKXD2YUX6XJ4TUZOOSV3DXQU6FTANCNFSM6AAAAAA2KFFVTA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Roberto Scotti
<https://scholar.google.it/citations?hl=en&user=EgZYcIAAAAAJ&view_op=list_works&sortby=pubdate>
 - Nuoro Forestry School
<https://www.facebook.com/NuoroForestrySchoolUniSS/>
Dip. Agraria <https://www.uniss.it/ugov/person/2496> - ***@***.*** ORCID
iD <https://orcid.org/0000-0001-7394-4473>  Publons
<https://publons.com/researcher/1104101/roberto-scotti/>   IRIS
<https://iris.uniss.it/cris/rp/rp02644>

-- 

--
*Dona il  5x1000* all'Università degli Studi di Sassaricodice fiscale: 
00196350904

**Reply by cderv** [reply]
> I could imagine but I do not know how I could find the answer you are
giving me.

It is not straightforward as **knitr** does not really allow passing computed variable to `label`. 

We document stuff like `knit_expand()` and child document in the R Markdown cookbook (which has **knitr** recipes which apply to quarto)
* https://bookdown.org/yihui/rmarkdown-cookbook/knit-expand.html
* https://bookdown.org/yihui/rmarkdown-cookbook/child-document.html
